### PR TITLE
feat: First view of a terminal UI for block and tx debugging.

### PR DESCRIPTION
**Pull request by DavidNix**
# Description, Motivation, and Context

Supports https://github.com/strangelove-ventures/ibctest/issues/123

I had to resort to some hackery to get basic subcommand support within the test binary.

## Demo

Sorry this is small. Right click and open in a new tab.

debug-tui1

## Road Map
* Tx detail view to inspect individual txs.
* Cosmos summary view that leverages existing db view v_cosmos_messages.
* Filter text when inspecting a tx to highlight and jump to text. 
* Live query view allowing you to type JSON path syntax to scope parts of txs. 
* Download txs as JSON for inspection with other tools. 

## Known Limitations, Trade-offs, Tech Debt
* It's really tough to test UIs but I have a test plan as the `tui.Model` becomes more stateful.

### CODE REVIEW

**Review by mark-rushakoff — APPROVED** [review]
At first I was wanting the terminal UI to be a separate binary, but shipping everything in one binary feels comfortable here. If we ever have a compelling reason to ship separately in the future, it will be easy enough to offer a migration path to users.

I somewhat wanted to see a couple more comments explaining what was happening in tui/views.go. It's difficult to follow exactly what is going on there at a quick glance, but it seems straightforward enough if you slow down to read it.

File: cmd/ibctest/ibctest_test.go [hunk 1]
```diff
@@ -202,4 +240,62 @@ func addFlags() {
...
+
+func runDebugTerminalUI(ctx context.Context) error {
+	dbPath := extraFlags.BlockDatabaseFile
+
+	// Explicitly check for file existence otherwise blockdb.ConnectDB implicitly creates and migrates a sqlite file.
+	if _, err := os.Stat(dbPath); err != nil {
+		return err
+	}
+
+	db, err := blockdb.ConnectDB(ctx, dbPath)
```
[the remark this answers is not in the document]
Comment (DavidNix) [reply cmd/ibctest/ibctest_test.go:271]: Great idea. I believe it's possible. Will investigate and add to a future PR if it's possible.

File: internal/blockdb/migrate.go [hunk 2]
```diff
@@ -173,11 +174,36 @@ SELECT
...
+	  , chain.id AS chain_kid
+	  , chain.chain_id AS chain_id
+	  , chain.chain_type AS chain_type
+	  , MAX(COALESCE(block.height, 0)) AS chain_height
+	  , COUNT(tx.data) AS tx_total
+    FROM test_case
+	LEFT JOIN chain ON chain.fk_test_id = test_case.id
+	LEFT JOIN block ON block.fk_chain_id = chain.id
+	LEFT JOIN tx ON tx.fk_block_id = block.id
+	GROUP BY test_case.id, chain.id
```
[the remark this answers is not in the document]
Comment (DavidNix) [reply internal/blockdb/migrate.go:201]: Hmmm, I thought my IDE always used spaces but I'll doublecheck and fix in a future PR.

File: internal/blockdb/query.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+type SchemaVersionResult struct {
+	GitSha string
+	// Always set to user's local time zone.
+	CreatedAt time.Time
+}
+
+// CurrentSchemaVersion returns the latest git sha and time that produced the sqlite schema.
+func (q *Query) CurrentSchemaVersion(ctx context.Context) (SchemaVersionResult, error) {
+	row := q.db.QueryRowContext(ctx, `SELECT git_sha, created_at FROM schema_version ORDER BY id DESC limit 1`)
```
[the remark this answers is not in the document]
Comment (DavidNix) [reply internal/blockdb/query.go:35]: I wish SQLite had a timestamp column. I felt it safer to let Go handle all the time transforming. I briefly explored `datetime` but didn't want to fight sqlite vs. Go's way of handling time. Additionally, my guess was time wasn't going to be a big part of the UI (but I may be be proven wrong there.)

**Comment by DavidNix** [comment]
> At first I was wanting the terminal UI to be a separate binary, but shipping everything in one binary feels comfortable here.

I also had the same thought. Until we automate releasing binaries through CI, I thought it less confusing to have it be a subcommand of `ibctest`.

**Review by agouin — APPROVED** [review]
Love the terminal UI and the way it's organized into the project

**Pull request merged**